clog/clog.go

```diff
@@ -0,0 +1,185 @@
+/*
+Package clog provides Context with logging information.
+*/
+package clog
+
+import (
+	"context"
+	"fmt"
+	"strconv"
+	"strings"
+	"sync"
+
+	"github.com/golang/glog"
+)
+
+// unique type to prevent assignment.
+type clogContextKeyT struct{}
+
+var clogContextKey = clogContextKeyT{}
+
+const (
+	// standard keys
+	manifestID    = "manifestID"
+	sessionID     = "sessionID"
+	nonce         = "nonce"
+	seqNo         = "seqNo"
+	orchSessionID = "orchSessionID" // session id generated on orchestrator for broadcaster
+)
+
+// Verbose is a boolean type that implements Infof (like Printf) etc.
+type Verbose bool
+
+var stdKeys map[string]bool
+var stdKeysOrder = []string{manifestID, sessionID, nonce, seqNo, orchSessionID}
+
+func init() {
+	stdKeys = make(map[string]bool)
+	for _, key := range stdKeysOrder {
+		stdKeys[key] = true
+	}
+}
+
+type values struct {
+	mu   sync.RWMutex
+	vals map[string]string
+}
+
+func newValues() *values {
+	return &values{
+		vals: make(map[string]string),
+	}
+}
+
+// Clone creates new context with parentCtx as parent and
+// logging details from logCtx
+func Clone(parentCtx, logCtx context.Context) context.Context {
+	cmap, _ := logCtx.Value(clogContextKey).(*values)
+	newCmap := newValues()
+	if cmap != nil {
+		cmap.mu.RLock()
+		for k, v := range cmap.vals {
+			newCmap.vals[k] = v
+		}
+		cmap.mu.RUnlock()
+	}
+	return context.WithValue(parentCtx, clogContextKey, newCmap)
+}
+
+func AddManifestID(ctx context.Context, val string) context.Context {
+	return AddVal(ctx, manifestID, val)
+}
+
+func AddSessionID(ctx context.Context, val string) context.Context {
+	return AddVal(ctx, sessionID, val)
+}
+
+func AddNonce(ctx context.Context, val uint64) context.Context {
+	return AddVal(ctx, nonce, strconv.FormatUint(val, 10))
+}
+
+func AddSeqNo(ctx context.Context, val uint64) context.Context {
+	return AddVal(ctx, seqNo, strconv.FormatUint(val, 10))
+}
+
+func AddOrchSessionID(ctx context.Context, val string) context.Context {
+	return AddVal(ctx, orchSessionID, val)
+}
+
+func AddVal(ctx context.Context, key, val string) context.Context {
+	cmap, _ := ctx.Value(clogContextKey).(*values)
+	if cmap == nil {
+		cmap = newValues()
+		ctx = context.WithValue(ctx, clogContextKey, cmap)
+	}
```

Comment:
Right now, after context is initialised, you can just do `clog.AddVal(ctx, "a", "b")` - this will work just fine because anyway it will return same context. I didn't want explicit initialisation because I didn't want to have accidental panic in production.